[E2E-SIM] 1 failures — 2026-04-16T22:07:34.255Z

## E2E Simulation Failure Report

**Run ID**: `c0c760b1-9d96-42db-abc0-91d41a794c67`
**Failed Checks**: 1/1
**Environment**: development

## Failed Assertions
- **user_refund:user[f3e2f424-8dd9-45df-8eb7-75291ad81ba0] refunds application[2e394d62-580a-42fb-a959-41cbbeb4f41a] for event[1e4a43d1-0357-4d9c-9c0e-0c8d2a0d9025]**: expected `"success"`, got `"status 200"`

## Axiom 디버깅
Edge Function 로그는 Axiom에서 확인할 수 있습니다.

```
axiom query "['edge-functions'] | where metadata.runId == 'c0c760b1-9d96-42db-abc0-91d41a794c67'"
```

또는 Axiom 대시보드에서 dataset `edge-functions`를 선택하고 `metadata.runId == 'c0c760b1-9d96-42db-abc0-91d41a794c67'`로 필터링하세요.

## Log Details
No storage log available

<details>
<summary>📋 Full Log (truncated to 60KB)</summary>

```log
[2026-04-16T22:07:06.699Z] [INFO] [tick:init] Found 5 E2E partners
[2026-04-16T22:07:07.082Z] [INFO] [tick:init] Selected 10/60 users for this tick
[2026-04-16T22:07:08.193Z] [WARN] [tick:partner_auth] No email found for partner 54435848-e269-44e3-9790-67309045246c
[2026-04-16T22:07:08.450Z] [WARN] [tick:partner_auth] No email found for partner 55032dc6-3aa5-4ef3-8c6b-0f9e964cd9d8
[2026-04-16T22:07:09.149Z] [WARN] [tick:partner_auth] No email found for partner c518ec3d-0988-4b47-9abc-6655763a45bb
[2026-04-16T22:07:09.411Z] [WARN] [tick:partner_auth] No email found for partner 2dae98b5-9f92-467b-9ad1-a590442f9cb6
[2026-04-16T22:07:09.657Z] [WARN] [tick:partner_auth] No email found for partner ae0f3970-6045-4a90-a9c0-5804fde1e2e1
[2026-04-16T22:07:30.883Z] [INFO] [tick:generate] Generated 1 actions
[2026-04-16T22:07:33.202Z] [ERROR] [tick:user_refund] user[f3e2f424-8dd9-45df-8eb7-75291ad81ba0] refunds application[2e394d62-580a-42fb-a959-41cbbeb4f41a] for event[1e4a43d1-0357-4d9c-9c0e-0c8d2a0d9025] → ❌
{"efAssertion":{"passed":true,"name":"refund_ef_ok","details":"status 200"},"dbAssertion":{"passed":false,"name":"refund_db_participant_removed","details":"event_participants row still exists for user=f3e2f424-8dd9-45df-8eb7-75291ad81ba0 event=1e4a43d1-0357-4d9c-9c0e-0c8d2a0d9025"}}
[2026-04-16T22:07:33.203Z] [ERROR] [tick:summary] Tick complete: 0/1 passed, 1 failed
```

</details>

--- a/supabase/functions/user-cancel-order/index.ts
+++ b/supabase/functions/user-cancel-order/index.ts
@@ -185,9 +185,11 @@ Deno.serve(withHandler(async (req) => {
         throw e;
       }
 
-      // Update DB: refund_status + refund_amount
+      // Update DB: status=cancelled + refund_status + refund_amount
+      // Fix #1515: Set status='cancelled' so on_application_cancel trigger removes event_participants
       const refundAmount = paymentAmount ?? (cancelResponse.amount as number | undefined);
       const updatePayload: Record<string, unknown> = {
+        status: "cancelled",
         refund_status: "completed",
         updated_at: new Date().toISOString(),
       };

--- a/supabase/migrations/20260417000002_remove_event_participants_on_cancel.sql
+++ b/supabase/migrations/20260417000002_remove_event_participants_on_cancel.sql
@@ -1,0 +1,37 @@
+-- Fix #1515: Remove event_participants when application is cancelled
+--
+-- INV-03 invariant: "Cancelled applications with orphaned participant records" was
+-- already tracked by the invariant monitor (db_invariant_monitor.sql), but there
+-- was no trigger to enforce cleanup.
+--
+-- When an event_application's status transitions to 'cancelled', the corresponding
+-- event_participants row must be removed to maintain DB consistency.
+-- The symmetric trigger (issue_ticket_on_approval) inserts the row on approval.
+
+set search_path to public, extensions;
+
+-- ────────────────────────────────────────────────────────
+-- Trigger function: delete event_participants on cancel
+-- ────────────────────────────────────────────────────────
+CREATE OR REPLACE FUNCTION public.remove_participant_on_cancel()
+RETURNS trigger
+LANGUAGE plpgsql
+SECURITY DEFINER
+SET search_path = public
+AS $$
+BEGIN
+  -- Only fire when status transitions TO 'cancelled'
+  IF NEW.status = 'cancelled' AND (OLD.status IS DISTINCT FROM 'cancelled') THEN
+    DELETE FROM public.event_participants
+    WHERE event_id = NEW.event_id
+      AND user_id = NEW.user_id;
+  END IF;
+  RETURN NEW;
+END;
+$$;
+
+DROP TRIGGER IF EXISTS on_application_cancel ON public.event_applications;
+
+CREATE TRIGGER on_application_cancel
+  AFTER UPDATE ON public.event_applications
+  FOR EACH ROW EXECUTE FUNCTION public.remove_participant_on_cancel();

--- a/supabase/migrations/20260417000002_remove_event_participants_on_cancel.sql
+++ b/supabase/migrations/20260417000002_remove_event_participants_on_cancel.sql
@@ -20,18 +20,20 @@ SECURITY DEFINER
 SET search_path = public
 AS $$
 BEGIN
-  -- Only fire when status transitions TO 'cancelled'
-  IF NEW.status = 'cancelled' AND (OLD.status IS DISTINCT FROM 'cancelled') THEN
-    DELETE FROM public.event_participants
-    WHERE event_id = NEW.event_id
-      AND user_id = NEW.user_id;
-  END IF;
+  DELETE FROM public.event_participants
+  WHERE event_id = NEW.event_id
+    AND user_id = NEW.user_id;
   RETURN NEW;
 END;
 $$;
 
 DROP TRIGGER IF EXISTS on_application_cancel ON public.event_applications;
 
+-- WHEN clause filters at the trigger level so remove_participant_on_cancel()
+-- is never invoked unless the row is actually transitioning to 'cancelled'.
+-- This avoids a function call on every event_applications UPDATE.
 CREATE TRIGGER on_application_cancel
   AFTER UPDATE ON public.event_applications
-  FOR EACH ROW EXECUTE FUNCTION public.remove_participant_on_cancel();
+  FOR EACH ROW
+  WHEN (NEW.status = 'cancelled' AND OLD.status IS DISTINCT FROM 'cancelled')
+  EXECUTE FUNCTION public.remove_participant_on_cancel();

--- a/supabase/tests/database/72_cancel_participant_cleanup_test.sql
+++ b/supabase/tests/database/72_cancel_participant_cleanup_test.sql
@@ -1,0 +1,165 @@
+-- Regression test for Fix #1515:
+-- Verifies on_application_cancel trigger correctly removes event_participants
+-- when an application status transitions to 'cancelled'.
+-- Symmetric to issue_ticket_on_approval which inserts on approval.
+--
+-- Cases:
+--   1. approved → cancelled : event_participants row deleted
+--   2. cancelled → cancelled: no-op (WHEN clause OLD.status IS DISTINCT FROM 'cancelled' guards re-fire)
+--   3. approved → rejected  : event_participants row preserved (trigger scoped to cancelled only)
+
+BEGIN;
+SELECT plan(4);
+
+SELECT tests.authenticate_as_service_role();
+
+-- ============================================================
+-- Shared result capture table
+-- ============================================================
+CREATE TEMP TABLE cancel_trigger_results (
+  test_case text,
+  participant_count int
+);
+
+-- ============================================================
+-- Case 1 + 2: approved → cancelled (deletion), then cancelled → cancelled (idempotency)
+-- ============================================================
+DO $$
+DECLARE
+  v_user_id    uuid;
+  v_partner_id uuid;
+  v_party_id   uuid;
+  v_event_id   uuid;
+  v_ticket_id  uuid;
+  v_app_id     uuid;
+BEGIN
+  v_user_id := tests.create_supabase_user('cancel_trigger_user_1');
+
+  INSERT INTO public.user_profiles (id, name, birth_date, gender, username)
+  VALUES (v_user_id, 'Cancel Test User 1', '2000-01-01', 'male', 'cancel_trigger_user_1')
+  ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name, birth_date = EXCLUDED.birth_date;
+
+  INSERT INTO public.partners (name)
+  VALUES ('Cancel Trigger Partner 1')
+  RETURNING id INTO v_partner_id;
+
+  INSERT INTO public.parties (partner_id, title, min_confirmed_count, max_participants)
+  VALUES (v_partner_id, 'Cancel Party 1', 0, 10)
+  RETURNING id INTO v_party_id;
+
+  INSERT INTO public.events (party_id, start_time, end_time, min_confirmed_count, max_participants)
+  VALUES (v_party_id, now() + interval '1 day', now() + interval '1 day 2 hours', 0, 10)
+  RETURNING id INTO v_event_id;
+
+  INSERT INTO public.tickets (event_id, name, quantity, price)
+  VALUES (v_event_id, 'Cancel Ticket 1', 10, 5000)
+  RETURNING id INTO v_ticket_id;
+
+  INSERT INTO public.event_applications (event_id, ticket_id, user_id, status)
+  VALUES (v_event_id, v_ticket_id, v_user_id, 'pending_review')
+  RETURNING id INTO v_app_id;
+
+  -- Approve: issue_ticket_on_approval trigger inserts into event_participants
+  UPDATE public.event_applications SET status = 'approved' WHERE id = v_app_id;
+
+  INSERT INTO cancel_trigger_results (test_case, participant_count)
+  SELECT 'before_cancel', count(*)::int
+  FROM public.event_participants
+  WHERE event_id = v_event_id AND user_id = v_user_id;
+
+  -- Cancel: on_application_cancel trigger should delete from event_participants
+  UPDATE public.event_applications SET status = 'cancelled' WHERE id = v_app_id;
+
+  INSERT INTO cancel_trigger_results (test_case, participant_count)
+  SELECT 'after_cancel', count(*)::int
+  FROM public.event_participants
+  WHERE event_id = v_event_id AND user_id = v_user_id;
+
+  -- Second cancel: WHEN (OLD.status IS DISTINCT FROM 'cancelled') prevents re-fire
+  UPDATE public.event_applications SET status = 'cancelled' WHERE id = v_app_id;
+
+  INSERT INTO cancel_trigger_results (test_case, participant_count)
+  SELECT 'after_second_cancel', count(*)::int
+  FROM public.event_participants
+  WHERE event_id = v_event_id AND user_id = v_user_id;
+END $$;
+
+-- Pre-condition: participant exists after approval
+SELECT is(
+  (SELECT participant_count FROM cancel_trigger_results WHERE test_case = 'before_cancel'),
+  1,
+  'Participant row inserted after approval (pre-condition for cancel test)'
+);
+
+-- Fix #1515: approved → cancelled deletes participant
+SELECT is(
+  (SELECT participant_count FROM cancel_trigger_results WHERE test_case = 'after_cancel'),
+  0,
+  'approved → cancelled: event_participants row deleted by on_application_cancel trigger'
+);
+
+-- Idempotency: cancelled → cancelled is no-op
+SELECT is(
+  (SELECT participant_count FROM cancel_trigger_results WHERE test_case = 'after_second_cancel'),
+  0,
+  'cancelled → cancelled: WHEN clause prevents re-fire, no error, count stays 0'
+);
+
+-- ============================================================
+-- Case 3: approved → rejected preserves event_participants
+-- ============================================================
+DO $$
+DECLARE
+  v_user_id    uuid;
+  v_partner_id uuid;
+  v_party_id   uuid;
+  v_event_id   uuid;
+  v_ticket_id  uuid;
+  v_app_id     uuid;
+BEGIN
+  v_user_id := tests.create_supabase_user('cancel_trigger_user_2');
+
+  INSERT INTO public.user_profiles (id, name, birth_date, gender, username)
+  VALUES (v_user_id, 'Cancel Test User 2', '2001-06-15', 'female', 'cancel_trigger_user_2')
+  ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name, birth_date = EXCLUDED.birth_date;
+
+  INSERT INTO public.partners (name)
+  VALUES ('Cancel Trigger Partner 2')
+  RETURNING id INTO v_partner_id;
+
+  INSERT INTO public.parties (partner_id, title, min_confirmed_count, max_participants)
+  VALUES (v_partner_id, 'Cancel Party 2', 0, 10)
+  RETURNING id INTO v_party_id;
+
+  INSERT INTO public.events (party_id, start_time, end_time, min_confirmed_count, max_participants)
+  VALUES (v_party_id, now() + interval '2 days', now() + interval '2 days 2 hours', 0, 10)
+  RETURNING id INTO v_event_id;
+
+  INSERT INTO public.tickets (event_id, name, quantity, price)
+  VALUES (v_event_id, 'Cancel Ticket 2', 10, 5000)
+  RETURNING id INTO v_ticket_id;
+
+  INSERT INTO public.event_applications (event_id, ticket_id, user_id, status)
+  VALUES (v_event_id, v_ticket_id, v_user_id, 'pending_review')
+  RETURNING id INTO v_app_id;
+
+  -- Approve: issue_ticket_on_approval inserts into event_participants
+  UPDATE public.event_applications SET status = 'approved' WHERE id = v_app_id;
+
+  -- Reject: on_application_cancel trigger WHEN clause (NEW.status = 'cancelled') does NOT fire
+  UPDATE public.event_applications SET status = 'rejected' WHERE id = v_app_id;
+
+  INSERT INTO cancel_trigger_results (test_case, participant_count)
+  SELECT 'after_reject', count(*)::int
+  FROM public.event_participants
+  WHERE event_id = v_event_id AND user_id = v_user_id;
+END $$;
+
+SELECT is(
+  (SELECT participant_count FROM cancel_trigger_results WHERE test_case = 'after_reject'),
+  1,
+  'approved → rejected: event_participants row preserved (trigger scoped to cancelled only)'
+);
+
+SELECT * FROM finish();
+ROLLBACK;

--- a/tests/backend_integration/src/scenario/refund_scenario_test.dart
+++ b/tests/backend_integration/src/scenario/refund_scenario_test.dart
@@ -496,5 +496,67 @@ void main() {
       expect(app['refund_status'], equals('none'),
           reason: 'refund_status must not change after ineligible refund attempt');
     });
+
+    test(
+        'S05-007: 취소 시 event_participants 삭제 — on_application_cancel 트리거 검증 (Fix #1515)',
+        () async {
+      // Regression test for #1515:
+      // INV-03: When event_applications.status → 'cancelled',
+      // on_application_cancel trigger must DELETE the event_participants row.
+
+      final ctx = await _createEvent(
+        fromNow: const Duration(days: 10),
+        price: 0,
+      );
+
+      // 1. Create application with status='approved' (triggers issue_ticket_on_approval → inserts participant)
+      final appRes = await adminClient
+          .from('event_applications')
+          .insert({
+            'event_id': ctx.eventId,
+            'ticket_id': ctx.ticketId,
+            'user_id': testUserId,
+            'status': 'approved',
+            'payment_amount': 0,
+            'refund_status': 'none',
+          })
+          .select()
+          .single();
+      final appId = appRes['id'] as String;
+
+      // 2. Wait for trigger: issue_ticket_on_approval inserts event_participants on status='approved'
+      final participantsBefore = await adminClient
+          .from('event_participants')
+          .select('id')
+          .eq('event_id', ctx.eventId)
+          .eq('user_id', testUserId);
+      expect(participantsBefore, isNotEmpty,
+          reason:
+              'Pre-condition: issue_ticket_on_approval should have inserted event_participants on approved');
+
+      // 3. Cancel the application (simulates what user-cancel-order EF does for free events)
+      await adminClient
+          .from('event_applications')
+          .update({'status': 'cancelled'}).eq('id', appId);
+
+      // 4. Verify: on_application_cancel trigger deleted the event_participants row
+      final participantsAfter = await adminClient
+          .from('event_participants')
+          .select('id')
+          .eq('event_id', ctx.eventId)
+          .eq('user_id', testUserId);
+      expect(participantsAfter, isEmpty,
+          reason:
+              'event_participants row must be deleted when application is cancelled (on_application_cancel trigger)');
+
+      // 5. Verify: application status is cancelled
+      final cancelledApp = await adminClient
+          .from('event_applications')
+          .select('status')
+          .eq('id', appId)
+          .single();
+      expect(cancelledApp['status'], equals('cancelled'),
+          reason: 'application status must be cancelled');
+    });
   });
 }

